src/expression/node/utils/assign.js

```diff
@@ -20,7 +20,11 @@ export function assignFactory ({ subset, matrix }) {
     try {
       if (Array.isArray(object)) {
         // we use matrix.subset here instead of the function subset because we must not clone the contents
-        return matrix(object).subset(index, value).valueOf()
+        const temp = matrix(object)
+        // make sure we do have the original array inside the temporary Matrix, we *want* to mutate it
+        temp._data = object
+        temp.subset(index, value)
+        return object
       } else if (object && typeof object.subset === 'function') { // Matrix
```

Comment:
I suppose another possibility would be to have a default non-aliasing (i.e., cloning) DenseMatrix constructor as well as an aliasing constructor.  Seems a little more complicated to me, but it could be workable while preserving encapsulation.

Reply:
This is definitely an ugly workaround. It was already there, and I only made it a bit more visible/understandable by introducing the `temp` variable and not using the return value of `subset`.

I've added a `FIXME` describing the workaround and a possible solution. I would like to focus this PR though on fixing the mutation issue in the `DenseMatrix` constructor and leave this workaround as it is for now.

Reply:
> It was already there,

To avoid mincing words, I completely disagree. Prior to this change, the DenseMatrix class behavior/interface was such that it created aliases (whether that was a good design decision or not), and the code here merely _used_ that property to get its work done. With this PR, the DenseMatrix design has fundamentally changed so that when relying on its public interface, DenseMatrix objects have an invariant that they never alias any object outside themselves. So now the code here goes around the public interface to modify internal data of a DenseMatrix object to violate that invariant. To me, that's an archetypical breach of encapsulation, and an invitation to maintenance mishaps when in the future someone modifies the code of DenseMatrix having no clue that there is one weird place in the code where the internal data **is** aliased.

I will go so far as to say that I would never merge code into a project of mine with such a violation, but of course I respect your prerogative to make a different decision here. But if so, I would _strongly_ recommend that you file an issue right away highlighting the problem, and hopefully someone can get to fixing it soon.

Reply:
Certainly that works, and doesn't violate any encapsulation, and if anyone encounters performance issues it could be fixed later to use a common assignment core with the subset method. I'd say go for it. Resolving this conversation.